File: nagasena/src/org/openexi/proc/io/ValueScannerLexical.java
package org.openexi.proc.io;

import java.io.IOException;
import java.io.InputStream;

import org.openexi.proc.common.QName;
import org.openexi.proc.grammars.Apparatus;
import org.openexi.schema.Characters;

final class ValueScannerLexical extends ValueScanner {

  private final ValueScanner m_baseValueScanner;
  private final StringValueScanner m_stringValueScanner;
  
  public ValueScannerLexical(ValueScanner baseValueScanner, StringValueScanner stringValueScanner) {
    m_baseValueScanner = baseValueScanner;
    m_stringValueScanner = stringValueScanner;
  }
  
  @Override
  public final QName getName() {
    return m_baseValueScanner.getName();
  }
  
  @Override
  public final short getCodecID() {
    return Apparatus.CODEC_LEXICAL;
  }

  @Override
  public int getBuiltinRCS(int simpleType) {
    return m_baseValueScanner.getBuiltinRCS(simpleType);
  }

  @Override
  public final void setInputStream(InputStream istream) {
    m_baseValueScanner.setInputStream(istream);
  }

  @Override
  public Characters scan(int name, int uri, int tp) throws IOException {
    return m_stringValueScanner.scan(name, uri, m_baseValueScanner.getBuiltinRCS(tp));
  }

}

File: nagasenacs/src/Nagasena/Proc/IO/ValueScannerLexical.cs
using System.IO;

using QName = Nagasena.Proc.Common.QName;
using Apparatus = Nagasena.Proc.Grammars.Apparatus;
using Characters = Nagasena.Schema.Characters;

namespace Nagasena.Proc.IO {

  internal sealed class ValueScannerLexical : ValueScanner {

    private readonly ValueScanner m_baseValueScanner;
    private readonly StringValueScanner m_stringValueScanner;

    public ValueScannerLexical(ValueScanner baseValueScanner, StringValueScanner stringValueScanner) {
      m_baseValueScanner = baseValueScanner;
      m_stringValueScanner = stringValueScanner;
    }

    public override QName Name {
      get {
        return m_baseValueScanner.Name;
      }
    }

    public override short CodecID {
      get {
        return Apparatus.CODEC_LEXICAL;
      }
    }

    public override int getBuiltinRCS(int simpleType) {
      return m_baseValueScanner.getBuiltinRCS(simpleType);
    }

    public override Stream InputStream {
      set {
        m_baseValueScanner.InputStream = value;
      }
    }

    public override Characters scan(int name, int uri, int tp) {
      return m_stringValueScanner.scan(name, uri, m_baseValueScanner.getBuiltinRCS(tp));
    }

  }

}